crates/astria-sequencer/src/app/proposal_fingerprint.rs

```diff
@@ -0,0 +1,192 @@
+use astria_eyre::eyre::{
+    bail,
+    Result,
+};
+use tendermint::{
+    abci,
+    Hash,
+};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) enum ProposalFingerprintData {
+    // No ProposalFingerprint has been set.
+    Unset,
+    // The set fingerprint is from a PrepareProposal request & response data.
+    Prepared([u8; 32]),
+    // The fingerprint from a checked PrepareProposal after a check & response data.
+    PreparedValid([u8; 32]),
+    // Fingerprint from an executed block, the cometbft block hash, if derived from previous
+    // prepare proposal, the second hash will be that previous value.
+    ExecutedBlock([u8; 32], Option<[u8; 32]>),
+    // The fingerprint failed comparison against a Prepared state, data is the previous
+    // fingerprint.
+    InvalidCheckedPrepared([u8; 32]),
```

Comment:
nit but prefer `CheckedPreparedMismatch` or something like that, it's a bit more descriptive as `Invalid` can mean the block was invalid, when it just means it wasn't executed